cli/src/commands/create/comments.rs

```diff
@@ -102,112 +140,235 @@ pub fn create(client: &Client, args: &CreateCommentsArgs) -> Result<()> {
+    let mut comments_to_put = Vec::with_capacity(batch_size);
+    let mut comments_to_sync = Vec::new();
+    let mut annotations = Vec::new();
 
-            if dataset_name.is_some() && new_comment.has_annotations() {
-                with_annotations.push((
-                    CommentUid(format!("{}.{}", source.id.0, new_comment.comment.id.0)),
-                    new_comment.clone(),
-                ));
-            }
-            batch.push(new_comment.comment);
+    // if --overwrite, everything will go to comments_to_sync, so put the default capacity there.
+    if overwrite {
+        std::mem::swap(&mut comments_to_put, &mut comments_to_sync)
+    }
+
+    let mut should_sync_comment = {
+        let mut seen = HashSet::new();
+        move |id: &CommentId| overwrite || (allow_duplicates && !seen.insert(id.clone()))
+    };
+
+    for read_comment_result in read_comments_iter(comments, Some(statistics)) {
+        let new_comment = read_comment_result?;
+
+        if dataset_name.is_some() && new_comment.has_annotations() {
+            annotations.push((
+                CommentUid(format!("{}.{}", source.id.0, new_comment.comment.id.0)),
+                new_comment.labelling,
+                new_comment.entities,
+            ));
         }
 
-        if batch.len() == batch_size || (!batch.is_empty() && eof) {
-            // Upload comments
-            let result = client
-                .sync_comments(&source.full_name(), &batch)
-                .chain_err(|| ErrorKind::Client("Could not upload batch of comments".into()))?;
-            let num_uploaded = batch.len();
-            statistics.add_comments(StatisticsUpdate {
-                uploaded: num_uploaded,
-                new: result.new,
-                updated: result.updated,
-                unchanged: result.unchanged,
-            });
-            batch.clear();
-
-            // Upload annotations
-            if let Some(dataset_name) = dataset_name.as_ref() {
-                for (comment_uid, new_comment) in with_annotations.iter() {
-                    client
-                        .update_labelling(
-                            &dataset_name,
-                            comment_uid,
-                            new_comment.labelling.as_ref(),
-                            new_comment.entities.as_ref(),
-                        )
-                        .chain_err(|| {
-                            ErrorKind::Client("Could not update labelling for comment".into())
-                        })?;
-                    statistics.add_annotation();
-                }
-                with_annotations.clear();
-            }
+        if should_sync_comment(&new_comment.comment.id) {
+            comments_to_sync.push(new_comment.comment);
+        } else {
+            comments_to_put.push(new_comment.comment);
         }
 
-        line_number += 1;
+        current_batch_size += 1;
+
+        if current_batch_size == batch_size {
```

Comment:
Looks like there's a problem with the batching logic here as `current_batch_size` is never reset.

Reply:
Good catch! `current_batch_size` was probably a premature optimization so I've taken it out in favour of always checking the sum of the two lengths.